### ISSUE TITLE
When will be the next release?

### ISSUE THREAD

**Issue by nikgraf** [ocaml/odoc#288]
There are changes in the master branch that would be really useful.

Do you have an idea/plan when to cut the next release?

**Comment by ostera** [comment]
cc @aantron

**Comment by pmetzger** [comment]
I'm a big believer in release early and often rather than when specific milestones are met, _unless_ functionality has significantly regressed.

**Comment by ostera** [comment]
I support @pmetzger — I’d like to see more releases as PRs get merged.

However, I understand that it takes time from maintainers to plan.

@aantron, what would it take to help out in cutting releases more often? More flexible milestones?

**Comment by aantron** [comment]
@ostera It doesn't take any significant time to plan, we just need to decide to do the release, which we can do now.

@pmetzger There are no specific milestones to "meet." As already said, that was a soft list, of things I thought would be "nice" to do for a release. They are not *blocking* the release. cc also @ryyppy about this.

I ask not to turn this issue into a generalized argument against non-flexible milestones, because we don't have the problem of non-flexible milestones in this repo in the first place.

If you want a release, just ask for one, and we'll see about making it. I just don't want to do releases at times when the only people using the latest `master` are active developers anyway, because it's work for no gain. That was the state the usage of odoc's `master` was in until a few weeks ago, so that's why there was no release. It has nothing to do with any kind of rigid planning.

I also don't want to convert a demand for the upcoming release, into a general rule to release often regardless of expressed demand or lack thereof, especially if the rule is based on a misunderstanding of the meaning of the repo using a GitHub "milestone" for prioritization. Again, if you want a release, just ask for it, and we will figure out what's going on at the time, and likely do a release.

So @ostera, @pmetzger, if you want the current `master`, specifically, released for your purposes, or know people that do, please say so, and again whenever this condition occurs in the future. Please encourage other people to do that as well :)

@ostera If you see a PR that is merged, that you would really like to be followed by a release, please comment on that PR, and encourage other people to do so as well. I would like to avoid an uncertain "rule" that merged PRs lead to releases, because obviously some shouldn't, others likely should, etc. I prefer to handle this kind of thing by ongoing communication.

It takes some time to actually *do* a release (rather than *plan* one), so it's good to limit the rate to at most bi-weekly or so, except when there are regressions (pretty rare), or when emergency bug fixes are needed.

@dbuenzli Do you have any objection to, or other comment on, a 1.3.1 release of `master` as it is now?

**Comment by dbuenzli** [comment]
I have partial fixes for #276, #272 and #282 in [this branch](https://github.com/dbuenzli/odoc/tree/fix-html-cmds). They at least avoid the segfaults I would be glad to have them in as this should allow me to push out a new version of odig later without too much hassle. 

I didn't PR yet because I wanted to have a look if I could fix them further -- i.e. implement the actual semantics rather than simply avoid the segfaults -- without investing too much time.

I think having @rizo's  #261 in would be nice too.

**Comment by nikgraf** [comment]
@aantron to provide you with more context why I was asking for: I planned to open PRs to add odoc to a couple of Reason projects. A lot of use BuckleScript bindings to external libraries e.g. ReasonReact and therefor without the change in how docs for `external` are done the API docs look confusing and complicated. A release with these changes would unblock me to move forward with that plan. That said I don't want to be pushy and can totally understand of it takes more time. In that case a rough timeline e.g. (2-4 more months till a release) would help me to plan.

**Comment by aantron** [comment]
Thanks @dbuenzli and @nikgraf. I suggest to wait for @dbuenzli's fixes, and I'll merge in #261, then we'll have a release.

@dbuenzli No pressure from odoc, however. If you need to defer working on any of those, please ping. We'll do a release without them, merge the fixes later, and do another release after that.

**Comment by ostera** [comment]
https://github.com/ocaml/odoc/pull/261 got merged, so, according to [this](https://github.com/ocaml/odoc/pull/292#issuecomment-461798110), we should be clear to release.

Unless I'm missing something — e.g, I see https://github.com/ocaml/odoc/issues/276 is still open.

![](https://media.giphy.com/media/lluj1cauAlO2vQEm8A/giphy.gif)

**Comment by aantron** [comment]
I think we should fix #304, 4.08 compatibility.

**Comment by aantron** [comment]
Ok, @jonludlam fixed #304 in #309 (thanks), so I will proceed with the release.

**Comment by dbuenzli** [comment]
@aantron could you please hold it a bit so that we test 4.08. I'm getting odoc assertions failures on the ocaml package with master.

**Comment by jonludlam** [comment]
Is it in Ctype.flatten_fields?

**Comment by dbuenzli** [comment]
Somewhere in `Ctype`: 

```
[FAILED][spawn:2207] ['/Users/dbuenzli/.opam/ocaml-variants.4.08.0+beta1/bin/odoc' 'compile' '--pkg' 'ocaml' '-o' '/tmp/odig-cache/odoc/ocaml/compiler-libs/CSE.odoc' '/Users/dbuenzli/.opam/ocaml-variants.4.08.0+beta1/lib/ocaml/compiler-libs/CSE.cmt' '-I' '/tmp/odig-cache/odoc/ocaml/' '-I' '/tmp/odig-cache/odoc/ocaml/compiler-libs/']:
 odoc: internal error, uncaught exception:
      File "typing/ctype.ml", line 327, characters 27-33: Assertion failed
      Raised at file "typing/ctype.ml", line 327, characters 27-39
      Called from file "src/loader/cmi.ml", line 328, characters 29-65
      Called from file "src/loader/cmt.ml", line 328, characters 4-42
      Called from file "src/loader/cmt.ml", line 347, characters 14-51
      Called from file "list.ml", line 121, characters 24-34
      Called from file "src/loader/cmt.ml", line 344, characters 2-343
      Called from file "src/loader/cmt.ml", line 515, characters 25-62
      Called from file "list.ml", line 121, characters 24-34
      Called from file "src/loader/cmt.ml", line 513, characters 4-136
      Called from file "src/loader/cmt.ml", line 522, characters 14-46
      Called from file "src/loader/odoc__loader.ml", line 137, characters 31-69
      Called from file "src/model/error.ml", line 65, characters 9-15
      Called from file "src/odoc/compile.ml", line 23, characters 8-36
      Called from file "cmdliner_term.ml", line 25, characters 19-24
      Called from file "cmdliner.ml", line 116, characters 32-39

```

**Comment by jonludlam** [comment]
Do you know what file it was processing?

**Comment by jonludlam** [comment]
Oh, I see

**Comment by aantron** [comment]
Yes. Also, I was planning to run at least `corpus.sh` on 4.08 as part of release work, though it may be largely redundant at this point.

**Comment by dbuenzli** [comment]
The answer to the question is today.

**Comment by aantron** [comment]
odoc 1.4.0 was [merged into opam](https://github.com/ocaml/opam-repository/pull/13617) a few hours ago, and is now installable after `opam update`. See the changelog [here](https://github.com/ocaml/odoc/releases/tag/1.4.0).

I'll make an announcement on discuss.ocaml.org shortly.

**Comment by pmetzger** [comment]
From here, would it make sense to have timed releases rather than feature based releases? Say every N months?

**Comment by aantron** [comment]
Yes, but we probably wouldn't commit to a specific time. But we would like to have releases more often.

Also, we didn't really have feature-based releases in any deliberate way. There were just a lot of bugs, so we wanted to get many of them fixed. As odoc gets more stable, and work becomes less intrusive (on average), releasing will become more normalized.

**Comment by aantron** [comment]
I made a note to consider releasing odoc again around May 1, in whatever (reasonable) state it is then. We can also do releases before that, if there are important bugfixes or features.
